README.md
# Skygear SDK for Android

Skygear Server is a cloud backend for making web and mobile app development
easier. [https://skygear.io](https://skygear.io)

The Skygear Android SDK library that gives you access to the Skygear Server
from your Android app.

## Getting Started

To get started, you need to have the
[Skygear Server](https://github.com/skygearIO/skygear-server) running and
Android SDK installed into your app. You can see detailed procedure at the
getting started guide at
[https://docs.skygear.io/android/guide](https://docs.skygear.io/guides/quickstart/android/).

You can sign up the Skygear Hosting at the Skygear Developer Portal at
[https://portal.skygear.io](https://portal.skygear.io)

Also, you can download the [sample application](https://docs.skygear.io/android/build/skygear-example-latest.apk)
to your Android devices to try it out.

## Documentation
The full documentation for Skygear is available on our docs site. The
[Android SDK get started guide](https://docs.skygear.io/guides/quickstart/android/) is a
good place to get started.


## Support

For implementation related questions or technical support, please find us on
the [official forum](https://discuss.skygear.io) or [community chat](https://slack.skygear.io).

If you believe you've found an issue with Skygear Android SDK, please feel free
to [report an issue](https://github.com/SkygearIO/skygear-SDK-Android/issues).


## How to contribute

Pull Requests Welcome!

We really want to see Skygear grows and thrives in the open source community.
If you have any fixes or suggestions, simply send us a pull request!

The current project setup assume Android Studio 3.0+. Download it here:
https://developer.android.com/studio/index.html and start your contribution!

## Continuous Delivery

Every release with version tagged will be push to
[jCenter](https://bintray.com/skygeario/maven/skygear-android/).


## License & Copyright

```
Copyright (c) 2015-present, Oursky Ltd.
All rights reserved.

This source code is licensed under the Apache License version 2.0
found in the LICENSE file in the root directory of this source tree.
An additional grant of patent rights can be found in the PATENTS
file in the same directory.

```


CHANGELOG.md
## 1.7.1 (2019-02-22)

### Bug Fixes

- Fix fail to call lambda without arguments #276

## 1.7.0 (2019-01-21)

### Features

- Support FCM (#254)

### Other Notes

- Update request timeout to 60s to align with iOS SDK


## 1.6.4 (2018-10-05)

### Other Notes

- Upgrade Java-WebSocket lib to 1.3.9
- Upgrade joda-time lib to 2.10

## 1.6.3 (2018-09-26)

### Features

- Support encrypt current user data when storing into SharedPreferences SkygearIO/features#250


## 1.6.2 (2018-07-23)

### Bug Fixes

- Add asset too large error code
- Add verification required error code

## 1.6.0 (2018-06-28)

### Features

- Add Lambda support for Skygear Data Type
- Auto save asset when saving record and lambda
- Implement joining queries with 'and'

### Other Notes

- Upgrade websocket lib to 1.3.8

## 1.5.0 (2018-04-23)

### Features

- Implement request verification and submit code

### Bug Fixes

- resetPassword API bug fix, change expireAt from Date to long
- Fix JSON serialization error of auth_data in signup and login api which
    happen in Android API version <= 18
- Should allow `JSONObject.NULL` in lambda arguments

### Other Notes

- Add admin prefix to disable/enable user methods
- Remove default endpoint in container
- Count support in query #211

## 1.4.1 (2018-03-09)

### Other Notes

- Revert "Upgrade gradle to version 4" for fixing missing dependencies

## 1.4.0 (2018-03-07)

### Features

- Support enable/disable user account

### Bug Fixes

- Fix potential memory leak

### Other Notes

- Update error code list
- Change func name and args for consistency for sso unlinkOAuthProvider api
- Upgrade gradle to version 4
- Supports various argument classes for callLambdaFunction
- Add checkstyle lint checking

## 1.3.1 (2018-02-01)

### Features

- Add ErrorSerializer
- Add SSO Plugin
    - Initial loginOAuthProvider (#150)
    - Link oauth provider (#168)
    - Unlink provider (#169)
    - Login and link with access token (#167)

### Bug Fixes

- Fix maybe null pointer exeception
- Serialise profile when signup

## 1.3.0 (2018-01-04)

### Features

- Add PubsubListener to hook up with PubsubClient onOpen, onClose and onError
  (#173)
- Support calling lambda with map (#120)
- Add loginWithCustomToken (#172)

### Bug Fixes

- Check null value when serializing and deserializing reserved fields

### Other Notes

- Reduce log on android test
- Change tools:context in activity_login layout

## 1.2.0 (2017-12-11)

### Bug Fixes

- Set AccessControl.Entry and AccessControl.Level public

## 1.1.1 (2017-10-23)

### Features

- Implement change password (#147)
- Implement save records atomically (#134)
- Add sdk version to request header

### Bug Fixes

- Fix example project login crash (#152)

### Other Notes

- Upgrade websocket lib (#165)
- Add support for Android SDK version 26 (#158)

## 1.1.0 (2017-08-07)

### Incompatible Changes

- Authentication-related function will return Record instead of User object

  In previous version of SKYKit, authentication methods return a User object
  which contains user-related information such as User ID, username and
  email. These information is moved to user record and the authentication
  methods are updated to return Record instead.

### Features

- Move user related role api to AuthContainer
- New signup login, remove user object (#142)

### Other Notes

- Upload doc prefixed with version to s3 bucket for CI

## 1.0.0 (2017-06-30)

### Incompatible Changes

- Update API grouping (#133, SkygearIO/features#71)

### Features

- Add forgot password methods to AuthContainer

### Other Notes

- Add copyright notice
- Update gradle build to 2.3.0 (#131)

## 0.24.1 (2017-05-23)

Revert "Move skygear to skygear_core module"

## 0.24.0 (2017-05-23)

### Features

- Set target SDK version to 25 (#130)

### Other Notes

- Move skygear to skygear_core module (#123)

## 0.23.1 (2017-05-02)

### Bug Fixes

- Remove dependency on jitpack.io (#124)

## 0.23.0 (2017-04-20)

### Features

- Return user friendly error message in error (#89)

### Other Notes

- Remove ignoring lint error on example project
- Change VolleyPlus to Volley

## 0.22.2 (2017-03-31)

### Features

- Add user query by username request and related methods
- Support content type for asset serialization (#92)

### Bug Fixes

- Fix the crash on parsing date string (#109)

### Other Notes

- Replaced StringBuffer with StringBuilder

## 0.22.1 (2017-02-24)

### Features

- Support limit and offset for record query (#98)

## 0.22.0 (2017-02-10)

No changes since last release.

## 0.21.0 (2017-01-11)

### Features

- Provide package name when register device (SkygearIO/skygear-server#239)

    Send package name as `topic` parameter when registering device, it will
    help cloud code to send push notification to segment of device.


## 0.20.0 (2016-12-20)

### Features

- Support signup anonymously (#83)
- Support unregister device (SkygearIO/skygear-server#245, SkygearIO/skygear-server#249)
- Implement Parcelable on Record class
- Implement UnknownValue class (skygeario/skygear-server#231)
- Support Push Notification (#67)

### Incompatible changes

- Add error code to request error (#75)

    It makes a breaking change to all `*ResponseHandler` changing the
    interface of fail callbacks

## 0.19.0 (2016-11-10)

No change since last release

## 0.18.0 (2016-10-26)

No change since last release

## 0.17.2 (2016-10-12)

### Features

- Allow setting `JSONObject.NULL` to record field (#61)
- Support transient include distance predicate (#56)

### Bug fixes

- Fix wss connection fail issue (#62)

## 0.17.1 (2016-10-04)

### Features

- Support distance related sorting predicates (#56)
- Support geo location (#56)
- Support transient and record reference (#49)
- Support asset (#43)


## 0.17.0 (2016-09-15)

### Features

- Add last login / last seen fields to user model (SkygearIO/skygear-server#110)

### Bug fixes

- Fix lambda function argument serialization issue (#46)


## 0.16.0 (2016-09-02)


skygear/src/main/java/io/skygear/skygear/Error.java
/*
 * Copyright 2017 Oursky Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package io.skygear.skygear;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * The Error for the response of Skygear Request.
 */
public class Error extends Exception {
    private final int codeValue;
    private final String name;
    private final String detailMessage;
    private final JSONObject info;

    /**
     * Instantiates a new Error.
     * <p>
     *     This constructor will creates an Error with
     *     error code {@link Code#UNEXPECTED_ERROR}
     * </p>
     *
     * @param detailMessage the detail message
     */
    public Error(String detailMessage) {
        this(Code.UNEXPECTED_ERROR.getValue(), null, detailMessage, null);
    }

    /**
     * Instantiates a new Error.
     * <p>
     *     This constructor will creates an Error and try to
     *     parse the error code value.
     * </p>
     *
     * @param codeValue     the code value
     * @param detailMessage the detail message
     */
    public Error(int codeValue, String detailMessage) {
        this(codeValue, null, detailMessage, null);
    }

    /**
     * Instantiates a new Error.
     *
     * @param codeValue     the code
     * @param name          the name of the error
     * @param detailMessage the detail message
     * @param info          the info from error message
     */
    public Error(int codeValue, String name, String detailMessage, JSONObject info) {
        super(Code.fromValue(codeValue).toString());

        this.codeValue = codeValue;
        this.name = name;
        this.detailMessage = detailMessage;
        this.info = info;
    }

    /**
     * Instantiates a new Error.
     *
     * @param jsonObject     the json error object
     * @throws JSONException if jsonObject does't contain error required attribute
     */
    public Error(JSONObject jsonObject) throws JSONException {
        this(jsonObject.getInt("code"), jsonObject.getString("name"), jsonObject.getString("message"), jsonObject.optJSONObject("info"));
    }

    /**
     * Gets error code.
     *
     * @return the code
     */
    public Code getCode() {
        return Code.fromValue(this.codeValue);
    }

    /**
     * The Error Codes.
     */
    public static enum Code {
        /**
         * Not authenticated error.
         */
        NOT_AUTHENTICATED(101),
        /**
         * Permission denied error.
         */
        PERMISSION_DENIED(102),
        /**
         * Access key not accepted error.
         */
        ACCESS_KEY_NOT_ACCEPTED(103),
        /**
         * Access token not accepted error.
         */
        ACCESS_TOKEN_NOT_ACCEPTED(104),
        /**
         * Invalid credentials error.
         */
        INVALID_CREDENTIALS(105),
        /**
         * Invalid signature error.
         */
        INVALID_SIGNATURE(106),
        /**
         * Bad request error.
         */
        BAD_REQUEST(107),
        /**
         * Invalid argument error.
         */
        INVALID_ARGUMENT(108),
        /**
         * Duplicated error.
         */
        DUPLICATED(109),
        /**
         * Resource not found error.
         */
        RESOURCE_NOT_FOUND(110),
        /**
         * Not supported error.
         */
        NOT_SUPPORTED(111),
        /**
         * Not implemented error.
         */
        NOT_IMPLEMENTED(112),
        /**
         * Constraint violated error.
         */
        CONSTRAINT_VIOLATED(113),
        /**
         * Incompatible schema error.
         */
        INCOMPATIBLE_SCHEMA(114),
        /**
         * Atomic operation failure error.
         */
        ATOMIC_OPERATION_FAILURE(115),
        /**
         * Partial operation failure error.
         */
        PARTIAL_OPERATION_FAILURE(116),
        /**
         * Undefined operation error.
         */
        UNDEFINED_OPERATION(117),
        /**
         * Plugin unavailable error.
         */
        PLUGIN_UNAVAILABLE(118),
        /**
         * Plugin timeout error.
         */
        PLUGIN_TIMEOUT(119),
        /**
         * Record query invalid error.
         */
        RECORD_QUERY_INVALID(120),
        /**
         * Plugin initializing error.
         */
        PLUGIN_INITIALIZING(121),
        /**
         * Response timeout error.
         */
        RESPONSE_TIMEOUT(122),
        /**
         * Denied argument error.
         */
        DENIED_ARGUMENT(123),
        /**
         * Record query denied error.
         */
        RECORD_QUERY_DENIED(124),
        /**
         * Not configured error.
         */
        NOT_CONFIGURED(125),
        /**
         * Password policy violated error.
         */
        PASSWORD_POLICY_VIOLATED(126),
        /**
         * User disabled error.
         */
        USER_DISABLED(127),
        /**
         * Verification required error.
         */
        VERIFICATION_REQUIRED(128),
        /**
         * Asset size too large error.
         */
        ASSET_SIZE_TOO_LARGE(129),
        /**
         * Unexpected error.
         */
        UNEXPECTED_ERROR(10000);

        private final int value;

        private Code(int value) {
            this.value = value;
        }

        /**
         * Gets value of the error code.
         *
         * @return the value
         */
        public int getValue() {
            return this.value;
        }

        /**
         * Creates error code from value.
         *
         * @param value the value
         * @return the code
         */
        public static Code fromValue(int value) {
            Code[] codes = Code.values();
            for (Code eachCode : codes) {
                if (eachCode.getValue() == value) {
                    return eachCode;
                }
            }

            return Code.UNEXPECTED_ERROR;
        }

        @Override
        public String toString() {
            switch (this) {
                case NOT_AUTHENTICATED:
                    return "You have to be authenticated to perform this operation.";
                case PERMISSION_DENIED:
                case ACCESS_KEY_NOT_ACCEPTED:
                case ACCESS_TOKEN_NOT_ACCEPTED:
                    return "You are not allowed to perform this operation.";
                case INVALID_CREDENTIALS:
                    return "You are not allowed to log in because the credentials you provided are not valid.";
                case INVALID_SIGNATURE:
                case BAD_REQUEST:
                    return "The server is unable to process the request.";
                case INVALID_ARGUMENT:
                    return "The server is unable to process the data.";
                case DUPLICATED:
                    return "This request contains duplicate of an existing resource on the server.";
                case RESOURCE_NOT_FOUND:
                    return "The requested resource is not found.";
                case NOT_SUPPORTED:
                    return "This operation is not supported.";
                case NOT_IMPLEMENTED:
                    return "This operation is not implemented.";
                case CONSTRAINT_VIOLATED:
                case INCOMPATIBLE_SCHEMA:
                case ATOMIC_OPERATION_FAILURE:
                case PARTIAL_OPERATION_FAILURE:
                    return "A problem occurred while processing this request.";
                case UNDEFINED_OPERATION:
                    return "The requested operation is not available.";
                case PLUGIN_INITIALIZING:
                case PLUGIN_UNAVAILABLE:
                    return "The server is not ready yet.";
                case PLUGIN_TIMEOUT:
                    return "The server took too long to process.";
                case RECORD_QUERY_INVALID:
                    return "The server is unable to process the query.";
                case RESPONSE_TIMEOUT:
                    return "The server timed out while processing the request.";
                case DENIED_ARGUMENT:
                    return "The server is unable to process the data.";
                case RECORD_QUERY_DENIED:
                    return "You are not allowed to perform this operation.";
                case NOT_CONFIGURED:
                    return "The server is not configured for this operation.";
                case PASSWORD_POLICY_VIOLATED:
                    return "The password does not meet policy requirement.";
                case USER_DISABLED:
                    return "The user is disabled.";
                case ASSET_SIZE_TOO_LARGE:
                    return "The asset is too large to upload.";
                case UNEXPECTED_ERROR:
                    return "An unexpected error has occurred.";
            }
            return super.toString();
        }
    }

    public String getName () {
        return this.name;
    }
    public int getCodeValue() {
        return this.codeValue;
    }

    public String getDetailMessage() {
        return this.detailMessage;
    }

    public JSONObject getInfo() {
        return this.info;
    }
}
